Src/Base/AMReX_IntVect.H

```diff
@@ -34,196 +35,235 @@ int coarsen (int i, int ratio) noexcept
 public:
+    static_assert(dim >= 1, "The number of dimensions of IntVectND must be positive");
 
-    class shift_hasher {
-    private:
-        static constexpr unsigned shift1 = sizeof(size_t)>=8 ? 20 : 10;
-        static constexpr unsigned shift2 = sizeof(size_t)>=8 ? 40 : 20;
-    public:
-        std::size_t operator()(const IntVect& vec) const noexcept
+    struct shift_hasher {
+        std::size_t operator()(const IntVectND<dim>& vec) const noexcept
         {
-            AMREX_D_DECL(std::size_t ret0 = vec[0], ret1 = vec[1], ret2 = vec[2]);
-#if AMREX_SPACEDIM == 1
-            return ret0;
-#elif AMREX_SPACEDIM == 2
-            return ret0 ^ (ret1 << shift1);
-#else
-            return ret0 ^ (ret1 << shift1) ^ (ret2 << shift2);
-#endif
+            static constexpr unsigned shift1 = sizeof(size_t)>=8 ? 20 : 10;
+            static constexpr unsigned shift2 = sizeof(size_t)>=8 ? 40 : 20;
+            if constexpr (dim == 1) {
+                return static_cast<std::size_t>(vec[0]);
+            } else if constexpr (dim == 2) {
+                return static_cast<std::size_t>(vec[0]) ^
+                      (static_cast<std::size_t>(vec[1]) << shift1);
+            } else if constexpr (dim == 3) {
+                return static_cast<std::size_t>(vec[0]) ^
+                      (static_cast<std::size_t>(vec[1]) << shift1) ^
+                      (static_cast<std::size_t>(vec[2]) << shift2);
+            } else {
+                std::size_t seed = dim;
+                // hash function from
+                // https://stackoverflow.com/questions/20511347/a-good-hash-function-for-a-vector
+                for (int i=0; i<dim; ++i) {
+                    auto x = static_cast<unsigned int>(vec[i]);
+                    x = ((x >> 16) ^ x) * 0x45d9f3b;
+                    x = ((x >> 16) ^ x) * 0x45d9f3b;
+                    x = (x >> 16) ^ x;
+                    seed ^= x + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+                }
+                return seed;
+            }
         }
     };
 
     using hasher = shift_hasher;
 
     ///
     /**
-       This is an IntVect all of whose components are equal to zero.
+    * \brief Construct an IntVectND whose components are all zero.
     */
-    static const IntVect Zero;
+    constexpr IntVectND () noexcept {} // cannot use = default due to Clang bug // NOLINT
 
-    ///
     /**
-       This is an IntVect all of whose components are equal to one.
+    * \brief Construct an IntVectND given the specific values for its
+    * coordinates. The inputs for this constructor are N integers,
+    * where N is equal to the number of dimensions of the IntVectND.
     */
-    static const IntVect Unit;
-
-    ///
-    /**
-    * \brief Construct an IntVect whose components are all zero.
-    */
-    constexpr IntVect () noexcept {} // cannot use = default due to Clang bug // NOLINT
+    template <class...Args,
+        std::enable_if_t<
+            (sizeof...(Args)+2 == dim) &&
```

Comment:
I am currently having trouble to get this constructor to work with
```C++
if constexpr (dim > 1) {
    py_iv.def(py::init<AMREX_D_DECL(int, int, int)>());
}
```
in pyAMReX.

Why `+2`? Shouldn't this be `-2` the way the signature is written below?

Reply:
Ah, cannot use the macro, obvs..